src/Rings/oscar_singular.jl

```diff
@@ -0,0 +1,448 @@
+"""
+iso_oscar_singular_coeff_ring
+
+"""
+    iso_oscar_singular_poly_ring(R::Ring, ...; kw...) -> Map
+
+Given a polynomial ring `R[x_1,...x_n]` return a ring isomorphism onto a native
+Singular ring `S[x_1,...,x_n]` with `R` isomorphic to `S`.
+"""
+iso_oscar_singular_poly_ring
+
+abstract type OscarSingularCoefficientRingMap{D, C} <: Map{D, C, Any, Any} end
+
+domain(f::OscarSingularCoefficientRingMap) = f.R
+
+codomain(f::OscarSingularCoefficientRingMap) = f.S
+
+# fallback
+function image(f::OscarSingularCoefficientRingMap, x)
+  parent(x) !== domain(f) && error("Element not in domain")
+  return codomain(f)(x)
+end
+
+(f::OscarSingularCoefficientRingMap)(x) = image(f, x)
+
+function preimage(f::OscarSingularCoefficientRingMap, x)
+  parent(x) !== codomain(f) && error("Element not in codomain")
+  return domain(f)(x)
+end
+
+# generic catchall 
+struct OscarSingularCoefficientRingMapGeneric{D,C} <: OscarSingularCoefficientRingMap{D,C} 
+  R::D
+  S::C
+  #OscarSingularCoefficientRingMapGeneric(R::D,S::C) where {D,C} = new{D,C}(R, S)
+end
+
+function iso_oscar_singular_coeff_ring(F::AbstractAlgebra.Ring)
+  return OscarSingularCoefficientRingMapGeneric(F, Singular.CoefficientRing(F))
+end
+
+# image(f, a) done by parent call overloading
+
+function preimage(f::OscarSingularCoefficientRingMapGeneric, a::Singular.n_unknown)
+  parent(a) !== codomain(f) && error("Element not in codomain")
+  b = Singular.libSingular.julia(Singular.libSingular.cast_number_to_void(a.ptr))
+  return b::elem_type(domain(f))
+end
+
+# ZZ
+iso_oscar_singular_coeff_ring(R::ZZRing) = OscarSingularCoefficientRingMapGeneric(R, Singular.Integers())
+
+# QQ
+iso_oscar_singular_coeff_ring(R::QQField) = OscarSingularCoefficientRingMapGeneric(R, Singular.Rationals())
+
+# prime field, small characteristic
+iso_oscar_singular_coeff_ring(R::fpField) = OscarSingularCoefficientRingMapGeneric(R, Singular.Fp(Int(characteristic(R))))
+
+# ZZ/nZZ, n small and big
+struct OscarSingularCoefficientRingMapZZMod{D} <: OscarSingularCoefficientRingMap{D, Singular.N_ZnRing}
+  R::D
+  S::Singular.N_ZnRing
+end
+
+iso_oscar_singular_coeff_ring(R::Union{zzModRing, ZZModRing}) = OscarSingularCoefficientRingMapZZMod(R, Singular.residue_ring(Singular.Integers(), BigInt(modulus(R)))[1]
+)
+
+# QQ(a)
+struct OscarSingularCoefficientRingMapAbsSimpleNumField <: OscarSingularCoefficientRingMap{AbsSimpleNumField, Singular.N_AlgExtField}
+  R::AbsSimpleNumField
+  S::Singular.N_AlgExtField
+end
+
+function iso_oscar_singular_coeff_ring(R::AbsSimpleNumField)
+  minpoly = defining_polynomial(R)
+  Qa = parent(minpoly)
+  a = gen(Qa)
+  SQa, (Sa,) = Singular.FunctionField(Singular.QQ, _variables_for_singular(symbols(Qa)))
+  Sminpoly = SQa(coeff(minpoly, 0))
+  for i in 1:degree(minpoly)
+    Sminpoly += SQa(coeff(minpoly, i))*Sa^i
+  end
+  SK, _ = Singular.AlgebraicExtensionField(SQa, Sminpoly)
+  return OscarSingularCoefficientRingMapAbsSimpleNumField(R, SK)
```

Comment:
Here I also wonder why you can't just use `OscarSingularCoefficientRingMapGeneric` ?
```suggestion
  return OscarSingularCoefficientRingMapGeneric(R, SK)
```